src/cdk/stepper/stepper.ts

```diff
@@ -0,0 +1,148 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {
+  ContentChildren,
+  EventEmitter,
+  Input,
+  Output,
+  QueryList,
+  Directive,
+  ViewChildren,
+  // tslint doesn't recognize `ElementRef` is used since it's only used as a generic.
+  // tslint:disable-next-line
+  ElementRef
+} from '@angular/core';
+import {CdkStep} from './step';
+import {LEFT_ARROW, RIGHT_ARROW, ENTER, SPACE} from '../keyboard/keycodes';
+import {coerceNumberProperty} from '../coercion/number-property';
+
+/** Used to generate unique ID for each stepper component. */
+let nextId = 0;
+
+/** Change event emitted on selection changes. */
+export class CdkStepperSelectionEvent {
+  /** The index of the step that is newly selected during this change event. */
+  newIndex: number;
+
+  /** The index of the step that was previously selected. */
+  oldIndex: number;
+
+  /** The step component that is selected ruing this change event. */
+  step: CdkStep;
+}
+
+@Directive({
+  selector: 'cdk-stepper',
+  host: {
+    '(focus)': '_setStepfocused()',
+    '(keydown)': '_onKeydown($event)',
+  },
+})
+export class CdkStepper {
+  /** The list of step components that the stepper is holding. */
+  @ContentChildren(CdkStep) _steps: QueryList<CdkStep>;
+
+  /** The list of step headers of the steps in the stepper. */
+  @ViewChildren('stepHeader') _stepHeader: QueryList<ElementRef>;
+
+  /** The index of the currently selected step. */
+  @Input()
+  get selectedIndex() { return this._selectedIndex; }
+  set selectedIndex(value: any) {
+    this._selectedIndex = coerceNumberProperty(value);
+  }
+  private _selectedIndex: number;
+
+  /** Event emitted when the selected step has changed. */
+  @Output() selectionChange = new EventEmitter<CdkStepperSelectionEvent>();
+
+  /** The index of the step that the focus is currently on. */
+  _focusIndex: number = 0;
+
+  private _groupId: number;
+
+  constructor() {
+    this._groupId = nextId++;
+  }
+
+  /** Selects and focuses the provided step. */
+  select(step: CdkStep | number): void {
```

Comment:
@mmalerba Would the `selected` still be returning the `this._selectedIndex` since the stepper cannot have the boolean `selected` attribute for each individual step?